File: internal/job/prose_test.go
package job

import (
	"encoding/json"
	"os"
	"testing"
)

// proseCase is one row of testdata/prose_cases.json, shared with the
// dashboard's JS twin so the two renderers cannot drift.
type proseCase struct {
	Name  string `json:"name"`
	Input string `json:"input"`
	Text  string `json:"text"`
	HTML  string `json:"html"`
}

func loadProseCases(t *testing.T) []proseCase {
	t.Helper()
	raw, err := os.ReadFile("testdata/prose_cases.json")
	if err != nil {
		t.Fatalf("read fixtures: %v", err)
	}
	var cases []proseCase
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatalf("parse fixtures: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("no fixture cases")
	}
	return cases
}

func TestRenderProseText_Fixtures(t *testing.T) {
	for _, c := range loadProseCases(t) {
		t.Run(c.Name, func(t *testing.T) {
			if got := RenderProseText(c.Input); got != c.Text {
				t.Errorf("\n got: %q\nwant: %q", got, c.Text)
			}
		})
	}
}

func TestParseProse_Structure(t *testing.T) {
	blocks := ParseProse("intro\n\n- a\n  - b\n\n```go\nx\n```")
	if len(blocks) != 3 {
		t.Fatalf("want 3 blocks, got %d: %+v", len(blocks), blocks)
	}
	if blocks[0].Kind != ProseParagraph || len(blocks[0].Lines) != 1 || blocks[0].Lines[0] != "intro" {
		t.Errorf("block 0: %+v", blocks[0])
	}
	list := blocks[1]
	if list.Kind != ProseList || list.Ordered || len(list.Items) != 1 {
		t.Fatalf("block 1: %+v", list)
	}
	item := list.Items[0]
	if len(item) != 2 || item[0].Kind != ProseParagraph || item[1].Kind != ProseList {
		t.Errorf("item blocks: %+v", item)
	}
	code := blocks[2]
	if code.Kind != ProseCode || code.Info != "go" || len(code.Lines) != 1 || code.Lines[0] != "x" {
		t.Errorf("block 2: %+v", code)
	}
}

func TestParseProse_OrderedStart(t *testing.T) {
	blocks := ParseProse("3. c\n4. d")
	if len(blocks) != 1 || !blocks[0].Ordered || blocks[0].Start != 3 || len(blocks[0].Items) != 2 {
		t.Fatalf("got %+v", blocks)
	}
}

func TestParseProse_HardBreakSplitsLines(t *testing.T) {
	blocks := ParseProse("a\\\nb  \nc\nd")
	if len(blocks) != 1 {
		t.Fatalf("got %+v", blocks)
	}
	want := []string{"a", "b", "c d"}
	got := blocks[0].Lines
	if len(got) != len(want) {
		t.Fatalf("lines: got %q want %q", got, want)
	}
	for i := range want {
		if got[i] != want[i] {
			t.Errorf("line %d: got %q want %q", i, got[i], want[i])
		}
	}
}

func TestRenderProseHTML_Fixtures(t *testing.T) {
	for _, c := range loadProseCases(t) {
		t.Run(c.Name, func(t *testing.T) {
			if got := RenderProseHTML(c.Input); got != c.HTML {
				t.Errorf("\n got: %q\nwant: %q", got, c.HTML)
			}
		})
	}
}

File: internal/web/jstest/prose.test.mjs
// Tests for internal/web/assets/js/prose.mjs — the client-side twin of
// internal/job/prose.go. Both read internal/job/testdata/prose_cases.json,
// and TestProseParity_GoAndJSAgree (handlers) runs this module from Go over
// the same inputs, so the two renderers cannot drift.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { renderProseHTML, parseProse } from "../assets/js/prose.mjs";

const here = dirname(fileURLToPath(import.meta.url));
const cases = JSON.parse(
  readFileSync(join(here, "..", "..", "job", "testdata", "prose_cases.json"), "utf8"),
);

test("fixtures are present", () => {
  assert.ok(cases.length > 0);
});

for (const c of cases) {
  test(`renderProseHTML: ${c.name}`, () => {
    assert.equal(renderProseHTML(c.input, c.links), c.html);
  });
}

test("renderProseHTML: inherited object properties are not link URLs", () => {
  // `links` is a plain object built by the caller; a bare `links[token]`
  // lookup would find Object.prototype.toString and link to a function.
  assert.equal(renderProseHTML("call toString twice", {}), "<p>call toString twice</p>");
  assert.equal(
    renderProseHTML("call `toString` twice", {}),
    "<p>call <code>toString</code> twice</p>",
  );
});

test("parseProse: structure of a mixed document", () => {
  const blocks = parseProse("intro\n\n- a\n  - b\n\n```go\nx\n```");
  assert.equal(blocks.length, 3);
  assert.deepEqual(blocks[0], { kind: "paragraph", lines: ["intro"] });
  assert.equal(blocks[1].kind, "list");
  assert.equal(blocks[1].ordered, false);
  assert.equal(blocks[1].items.length, 1);
  assert.equal(blocks[1].items[0][1].kind, "list");
  assert.deepEqual(blocks[2], { kind: "code", fence: "```", info: "go", lines: ["x"] });
});

test("renderProseHTML: null and undefined render empty", () => {
  assert.equal(renderProseHTML(null), "");
  assert.equal(renderProseHTML(undefined), "");
});

test("renderProseHTML: a missing links argument links nothing", () => {
  assert.equal(renderProseHTML("see mk4Qz1"), "<p>see mk4Qz1</p>");
});
